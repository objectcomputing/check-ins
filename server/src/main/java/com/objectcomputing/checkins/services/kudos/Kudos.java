package com.objectcomputing.checkins.services.kudos;

import io.micronaut.core.annotation.Introspected;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.data.annotation.AutoPopulated;
import io.micronaut.data.annotation.DateCreated;
import io.micronaut.data.annotation.TypeDef;
import io.micronaut.data.annotation.sql.ColumnTransformer;
import io.micronaut.data.model.DataType;
import io.swagger.v3.oas.annotations.media.Schema;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

import java.time.LocalDate;
import java.util.Objects;
import java.util.UUID;

@Entity
@Introspected
@Table(name = "kudos")
public class Kudos {

    @Id
    @Column(name = "id")
    @AutoPopulated
    @TypeDef(type = DataType.STRING)
    @Schema(description = "the id of the kudos", required = true)
    private UUID id;

    @NotBlank
    @Column(name = "message")
    @ColumnTransformer(read = "pgp_sym_decrypt(message::bytea, '${aes.key}')", write = "pgp_sym_encrypt(?, '${aes.key}')")
    @Schema(description = "message describing the kudos", required = true)
    private String message;

    @Nullable
    @Column(name = "publiclyvisible")
    @Schema(description = "true if the kudos is public", required = true)
    private Boolean publiclyVisible;

    @NotNull
    @Column(name = "senderid")
    @TypeDef(type = DataType.STRING)
    @Schema(description = "id of the user who gave the kudos", required = true)
    private UUID senderId;

    @Nullable
    @Column(name = "teamid")
    @TypeDef(type = DataType.STRING)
    @Schema(description = "id of the team that received the kudos, null if anonymous group of members")
    private UUID teamId;

    @Column(name = "datecreated")
    @DateCreated
    @TypeDef(type = DataType.DATE)
    @Schema(description = "date the kudos were created", required = true)
    private LocalDate dateCreated;

    @Nullable
    @Column(name = "dateapproved")
    @TypeDef(type = DataType.DATE)
    @Schema(description = "date the kudos were approved, null if pending")
    private LocalDate dateApproved;

    public Kudos() {}

    /**
     * Constructor for creating Kudos
     * @param message string describing the kudos
     * @param senderId id of the user who gave the kudos
     */
    public Kudos(String message, UUID senderId, Boolean publiclyVisible) {
        this.message = message;
        this.senderId = senderId;
        this.dateApproved = null;
        this.publiclyVisible = publiclyVisible;
    }

    /**
     * Constructor for updating Kudos
     * @param id the id of the kudos
     * @param message string describing the kudos
     * @param senderId id of the user who gave the kudos
     * @param dateApproved date the kudos were approved
     */
    public Kudos(UUID id, String message, UUID senderId, @Nullable LocalDate dateApproved, Boolean publiclyVisible) {
        this.id = id;
        this.message = message;
        this.senderId = senderId;
        this.dateApproved = dateApproved;
        this.publiclyVisible = publiclyVisible;
    }

    public Kudos(KudosCreateDTO kudosCreateDTO) {
        this.message = kudosCreateDTO.getMessage();
        this.senderId = kudosCreateDTO.getSenderId();
        this.teamId = kudosCreateDTO.getTeamId();
        this.publiclyVisible = kudosCreateDTO.getPubliclyVisible();
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public UUID getSenderId() {
        return senderId;
    }

    public void setSenderId(UUID senderId) {
        this.senderId = senderId;
    }

    public Boolean getPubliclyVisible() {
        return publiclyVisible;
    }

    public void setPubliclyVisible(Boolean publiclyVisible) {
        this.publiclyVisible = publiclyVisible;
    }

    @Nullable
    public UUID getTeamId() {
        return teamId;
    }

    public void setTeamId(@Nullable UUID teamId) {
        this.teamId = teamId;
    }

    public LocalDate getDateCreated() {
        return dateCreated;
    }

    public void setDateCreated(LocalDate dateCreated) {
        this.dateCreated = dateCreated;
    }

    @Nullable
    public LocalDate getDateApproved() {
        return dateApproved;
    }

    public void setDateApproved(@Nullable LocalDate dateApproved) {
        this.dateApproved = dateApproved;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Kudos kudos = (Kudos) o;
        return Objects.equals(id, kudos.id) && Objects.equals(message, kudos.message) && Objects.equals(publiclyVisible, kudos.publiclyVisible) && Objects.equals(senderId, kudos.senderId) && Objects.equals(dateCreated, kudos.dateCreated) && Objects.equals(dateApproved, kudos.dateApproved);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, message, senderId, dateCreated, dateApproved, publiclyVisible);
    }   

    @Override
    public String toString() {
        return "Kudos{" +
                "id=" + id +
                ", message='" + message + '\'' +
                ", senderId=" + senderId +
                ", dateCreated=" + dateCreated +
                ", dateApproved=" + dateApproved +
                ", publiclyVisible=" + publiclyVisible +
                '}';
    }
}
