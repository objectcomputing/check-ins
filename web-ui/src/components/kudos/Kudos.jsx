import { Box, Card, CardContent, CardHeader, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import React from 'react'
import AvatarComponent from '../avatar/Avatar'

const PREFIX = 'Kudos';
const classes = {
  display: `${PREFIX}-display`,
  small: `${PREFIX}-small`,
  large: `${PREFIX}-large`
};

const StyledCard = styled(Card)(({theme}) => ({
  [`& .${classes.display}`]: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "space-around"
  },
  [`& .${classes.small}`]: {
    width: theme.spacing(6),
    height: theme.spacing(6),
    margin: theme.spacing(1),
  },
  [`& .${classes.large}`]: {
    width: theme.spacing(8),
    height: theme.spacing(8),
  }
}));

export default function Kudos({kudosTo, kudosFrom, content}) {
  return (
    <StyledCard classes={{root: classes.display}}>
      <Box display="flex" justifyContent="center">
        <CardHeader
          avatar={
            <AvatarComponent className={classes.large} imageUrl={kudosTo.imageUrl} />
          }
          disableTypography
          title={
            <Typography variant="h5" component="h2">
              {`Kudos to ${kudosTo.name} ðŸŽ‰`}
            </Typography>
          }
        />
      </Box>
      <CardContent>
      <Typography>
        <Box component={'span'} textAlign="center">
            {content}
        </Box>
      </Typography>
      {kudosFrom && 
        <Box display="flex" justifyContent="flex-end" alignItems="center">
          <Typography variant="subtitle2">
              {`${kudosFrom.name} - ${kudosFrom.title}`}
            </Typography>
          <AvatarComponent className={classes.small} imageUrl={kudosFrom.imageUrl} />
        </Box>
      }
      </CardContent>
    </StyledCard>
  );
}


