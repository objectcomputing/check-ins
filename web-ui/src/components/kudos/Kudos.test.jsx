import React from 'react';
import renderer from 'react-test-renderer';
import Kudos from './Kudos';
import image from '../../logo.svg';

const kudosData = {
  kudosTo: {
    name: 'John Doe',
    imageUrl: image
  },
  content:
    'Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.',
  kudosFrom: {
    name: 'Jane Doe',
    imageUrl: image,
    title: 'Senior Software Engineer'
  }
};

it("renders correctly when 'kudosFrom' prop is provided", () => {
  const component = renderer.create(
    <Kudos
      kudosTo={kudosData.kudosTo}
      content={kudosData.content}
      kudosFrom={kudosData.kudosFrom}
    />
  );
  expect(component.toJSON()).toMatchSnapshot();
});

it("renders correctly when 'kudosFrom' prop is not provided", () => {
  const component = renderer.create(
    <Kudos kudosTo={kudosData.kudosTo} content={kudosData.content} />
  );
  expect(component.toJSON()).toMatchSnapshot();
});
