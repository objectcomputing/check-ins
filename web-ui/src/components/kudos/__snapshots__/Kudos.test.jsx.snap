// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly when 'kudosFrom' prop is not provided 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root Kudos-display css-3q8kj4-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiBox-root css-1l4w6pd"
  >
    <div
      className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        className="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular Kudos-large css-1wlk0hk-MuiAvatar-root"
        >
          <img
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/src/logo.svg"
          />
        </div>
      </div>
      <div
        className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <h2
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Kudos to John Doe ðŸŽ‰
        </h2>
      </div>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-xi606m"
      >
        Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
      </span>
    </p>
  </div>
</div>
`;

exports[`renders correctly when 'kudosFrom' prop is provided 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root Kudos-display css-3q8kj4-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiBox-root css-1l4w6pd"
  >
    <div
      className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        className="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular Kudos-large css-1wlk0hk-MuiAvatar-root"
        >
          <img
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/src/logo.svg"
          />
        </div>
      </div>
      <div
        className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <h2
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Kudos to John Doe ðŸŽ‰
        </h2>
      </div>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      <span
        className="MuiBox-root css-xi606m"
      >
        Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
      </span>
    </p>
    <div
      className="MuiBox-root css-pori7h"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
      >
        Jane Doe - Senior Software Engineer
      </h6>
      <div
        className="MuiAvatar-root MuiAvatar-circular Kudos-small css-1wlk0hk-MuiAvatar-root"
      >
        <img
          className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="/src/logo.svg"
        />
      </div>
    </div>
  </div>
</div>
`;
